lib/pdn/pdn.c

```diff
@@ -0,0 +1,763 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include "pdn.h"
+#include "at_cmd.h"
+#include "at_notif.h"
+
+#include <errno.h>
+#include <string.h>
+#include <stdio.h>
+#include <net/socket.h>
+
+/**< Maximum buffer size. */
+#define PDN_AT_BUFFER_SIZE 256
+
+#define PDN_MAX_APN_NAME 100
+#define PDN_ID_UNINITIALIZED 0xFFFFFFFF
+
+/** @brief PDN states
+ * Internal definitions to track PDN state
+ */
+enum pdn_conn_state {
+	/**< PDN is idle/unused. */
+	PDN_IDLE = 0,
+	/**< PDN is allocated.   */
+	PDN_ALLOCATED = 1,
+	/**< PDN is created, has an assigned context identifier.   */
+	PDN_ASSIGNED = 2,
+	/**< PDN is connected to an APN and activated. */
+	PDN_ACTIVE = 3,
+	/**< PDN is being deactivated or disconnected from the APN. */
+	PDN_CLOSING = 4,
+};
+
+/** @brief PDN type
+ * Internal definitions to track PDN inet types
+ */
+enum { PDN_AF_INET = 1, PDN_AF_INET6 = 1 << 1 };
+
+/** @brief AT commands
+ * Internal definitions for supported AT command
+ */
+enum pdn_management_at_cmd {
+	CMD_CONTEXT_CREATE = 0,
+	CMD_APN_CONNECT,
+	CMD_PDN_ACTIVATE,
+	CMD_PDN_ID_GET,
+	CMD_PDN_STATE_GET,
+	CMD_PDN_DEACTIVATE,
+	CMD_APN_DISCONNECT,
+	CMD_MAX
+};
+
+/** @brief AT command responses
+ * Internal definitions for supported AT command responses
+ */
+enum pdn_management_at_evt {
+	EVT_OK = 0,
+	EVT_CONTEXT_CREATE,
+	EVT_PDN_ID_GET,
+	EVT_PDN_STATE_GET,
+	EVT_CONTEXT_QUERY,
+	EVT_MAX
+};
+
+/** Data type for PDN connection. */
+struct pdn_connection {
+	/* Client defined callback when PDN related nofication is received */
+	pdn_notif_handler_t callback;
+	/* Modem assigned identifier for uniquely identifying a PDN. */
+	int context_id;
+	/* PDN id. */
+	unsigned int pdn_id;
+	/* Indicates the connection state #pdn_conn_state of the PDN. */
+	volatile unsigned int state;
```

Comment:
There should be no need for `volatile` here.

Reply:
removed